immer/detail/type_traits.hpp

```diff
@@ -0,0 +1,158 @@
+//
+// immer - immutable data structures for C++
+// Copyright (C) 2016, 2017 Juan Pedro Bolivar Puente
+//
+// This file is part of immer.
+//
+// immer is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// immer is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with immer.  If not, see <http://www.gnu.org/licenses/>.
+//
+
+#pragma once
+
+#include <type_traits>
+#include <iterator>
+#include <utility>
+
+namespace immer {
+namespace detail {
+
+template <typename... Ts>
+struct make_void { using type = void; };
+
+template <typename... Ts>
+using void_t = typename make_void<Ts...>::type;
+
+template <typename T, typename = void>
+struct is_dereferenceable_t : std::false_type {};
+
+template <typename T>
+struct is_dereferenceable_t<T, void_t<decltype(*(std::declval<T&>()))>> :
+    std::true_type {};
+
+template <typename T>
+constexpr bool is_dereferenceable = is_dereferenceable_t<T>::value;
```

Comment:
I would suggest using the same naming conventions as in the `std` for traits:
1. The "naked" trait has no suffix
2. The `::type` alias has `_t` suffix
3. The `::value` or `::type::value` alias has `_v` suffix.

This means you should rename `is_deferenceable_t` as `is_deferenceable`, and rename the current `is_dereferenceable` as `is_dereferenceable_v`.

Reply:
Will do!

Reply:
Thanks! :+1: